File: Sep12/Lecture7Examples/Example26/MainActivity.java
package com.varunshankar.example26;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.os.HandlerCompat;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import java.net.URL;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import org.json.JSONException;

public class MainActivity extends AppCompatActivity
                        implements View.OnClickListener{

    private EditText mEtLocation;
    private TextView mTvTemp;
    private TextView mTvPress;
    private TextView mTvHum;
    private WeatherData mWeatherData;
    private Button mBtSubmit;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //Get the edit text and all the text views
        mEtLocation = (EditText) findViewById(R.id.et_location);
        mTvTemp = (TextView) findViewById(R.id.tv_temp);
        mTvPress = (TextView) findViewById(R.id.tv_pressure);
        mTvHum = (TextView) findViewById(R.id.tv_humidity);

        mBtSubmit = (Button) findViewById(R.id.button_submit);
        mBtSubmit.setOnClickListener(this);
    }
    @Override
    public void onClick(View view) {
        switch(view.getId()){
            case R.id.button_submit:{
                //Get the string from the edit text and sanitize the input
                String inputFromEt = mEtLocation.getText().toString().replace(' ','&');
                loadWeatherData(inputFromEt);
            }
            break;
        }
    }

    private void loadWeatherData(String location){
        new FetchWeatherTask().execute(location);
    }

    private class FetchWeatherTask{
       ExecutorService executorService = Executors.newSingleThreadExecutor();
       Handler mainThreadHandler = HandlerCompat.createAsync(Looper.getMainLooper());
       public void execute(String location)
       {
            executorService.execute(new Runnable(){
                @Override
                public void run(){
                    String jsonWeatherData;
                    URL weatherDataURL = NetworkUtils.buildURLFromString(location);
                    jsonWeatherData = null;
                    try{
                        jsonWeatherData = NetworkUtils.getDataFromURL(weatherDataURL);
                        postToMainThread(jsonWeatherData);
                    }catch(Exception e){
                        e.printStackTrace();
                    }
                }
            });
       }

       private void postToMainThread(String jsonWeatherData)
       {
           mainThreadHandler.post(new Runnable() {
               @Override
               public void run() {
                   if (jsonWeatherData != null) {
                       try {
                           mWeatherData = JSONWeatherUtils.getWeatherData(jsonWeatherData);
                       } catch (JSONException e) {
                           e.printStackTrace();
                       }
                       if (mWeatherData != null) {
                           mTvTemp.setText("" + Math.round(mWeatherData.getTemperature().getTemp() - 273.15) + " C");
                           mTvHum.setText("" + mWeatherData.getCurrentCondition().getHumidity() + "%");
                           mTvPress.setText("" + mWeatherData.getCurrentCondition().getPressure() + " hPa");
                       }
                   }
               }
           });
       }
    }

/*    private class FetchWeatherTask extends AsyncTask<String,Void,String> {
        @Override
        protected String doInBackground(String... inputStringArray) {
            String location = inputStringArray[0];
            URL weatherDataURL = NetworkUtils.buildURLFromString(location);
            String jsonWeatherData = null;
            try{
                jsonWeatherData = NetworkUtils.getDataFromURL(weatherDataURL);
                return jsonWeatherData;
            }catch(Exception e){
                e.printStackTrace();
                return null;
            }
        }

        @Override
        protected void onPostExecute(String jsonWeatherData) {
            if (jsonWeatherData!=null){
                try {
                    mWeatherData = JSONWeatherUtils.getWeatherData(jsonWeatherData);
                }
                catch(JSONException e){
                    e.printStackTrace();
                }
                if(mWeatherData!=null) {
                    mTvTemp.setText("" + Math.round(mWeatherData.getTemperature().getTemp() - 273.15) + " C");
                    mTvHum.setText("" + mWeatherData.getCurrentCondition().getHumidity() + "%");
                    mTvPress.setText("" + mWeatherData.getCurrentCondition().getPressure() + " hPa");
                }
            }
        }
    }*/
}

File: Sep12/Example26/app/src/main/java/com/varunshankar/example26/MainActivity.kt
package com.varunshankar.example26

import androidx.appcompat.app.AppCompatActivity
import android.widget.EditText
import android.widget.TextView
import android.os.Bundle
import androidx.core.os.HandlerCompat
import android.os.Looper
import android.view.View
import android.widget.Button
import org.json.JSONException
import java.lang.Exception
import java.util.concurrent.Executors
import kotlin.math.roundToInt

class MainActivity : AppCompatActivity(), View.OnClickListener {
    private var mEtLocation: EditText? = null
    private var mTvTemp: TextView? = null
    private var mTvPress: TextView? = null
    private var mTvHum: TextView? = null
    private var mWeatherData: WeatherData? = null
    private var mBtSubmit: Button? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //Get the edit text and all the text views
        mEtLocation = findViewById<View>(R.id.et_location) as EditText
        mTvTemp = findViewById<View>(R.id.tv_temp) as TextView
        mTvPress = findViewById<View>(R.id.tv_pressure) as TextView
        mTvHum = findViewById<View>(R.id.tv_humidity) as TextView
        mBtSubmit = findViewById<View>(R.id.button_submit) as Button
        mBtSubmit!!.setOnClickListener(this)
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.button_submit -> {

                //Get the string from the edit text and sanitize the input
                val inputFromEt = mEtLocation!!.text.toString().replace(' ', '&')
                loadWeatherData(inputFromEt)
            }
        }
    }

    private fun loadWeatherData(location: String) {
        FetchWeatherTask().execute(location)
    }

    private inner class FetchWeatherTask {
        var executorService = Executors.newSingleThreadExecutor()!!
        var mainThreadHandler = HandlerCompat.createAsync(Looper.getMainLooper())
        fun execute(location: String?) {
            executorService.execute {
                var jsonWeatherData: String?
                val weatherDataURL = NetworkUtils.buildURLFromString(location!!)
                try {
                    jsonWeatherData = NetworkUtils.getDataFromURL(weatherDataURL!!)
                    postToMainThread(jsonWeatherData)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }

        private fun postToMainThread(jsonWeatherData: String?) {
            mainThreadHandler.post {
                if (jsonWeatherData != null) {
                    try {
                        mWeatherData = JSONWeatherUtils.getWeatherData(jsonWeatherData)
                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }
                    if (mWeatherData != null) {
                        mTvTemp!!.text =
                            "" + (mWeatherData!!.temperature.temp - 273.15).roundToInt() + " C"
                        mTvHum!!.text = "" + mWeatherData!!.currentCondition.humidity + "%"
                        mTvPress!!.text = "" + mWeatherData!!.currentCondition.pressure + " hPa"
                    }
                }
            }
        }
    }
// Use this to discuss AsyncTask, mention its deprecation. Don't talk about leaks just yet.
/*private class FetchWeatherTask extends AsyncTask<String,Void,String> {
        @Override
        protected String doInBackground(String... inputStringArray) {
            String location = inputStringArray[0];
            URL weatherDataURL = NetworkUtils.buildURLFromString(location);
            String jsonWeatherData = null;
            try{
                jsonWeatherData = NetworkUtils.getDataFromURL(weatherDataURL);
                return jsonWeatherData;
            }catch(Exception e){
                e.printStackTrace();
                return null;
            }
        }

        @Override
        protected void onPostExecute(String jsonWeatherData) {
            if (jsonWeatherData!=null){
                try {
                    mWeatherData = JSONWeatherUtils.getWeatherData(jsonWeatherData);
                }
                catch(JSONException e){
                    e.printStackTrace();
                }
                if(mWeatherData!=null) {
                    mTvTemp.setText("" + Math.round(mWeatherData.getTemperature().getTemp() - 273.15) + " C");
                    mTvHum.setText("" + mWeatherData.getCurrentCondition().getHumidity() + "%");
                    mTvPress.setText("" + mWeatherData.getCurrentCondition().getPressure() + " hPa");
                }
            }
        }
    }*/
}